### PR TITLE
BAI-220: S3 HEAD validation in orchestrator and NDJSON reader

### PR DESCRIPTION
**Pull request by adachaman-11**
## Summary
- **`$import` endpoint** (`import.js`): validates S3 URIs and enforces a fail-closed bucket allowlist (`BULK_IMPORT_ALLOWED_S3_BUCKETS`) before creating the Task — an empty allowlist rejects all requests. Kafka publish is wrapped in try/catch; on failure, the Task is marked `failed` via `markTaskFailedAsync` to prevent orphaned tasks.
- **Orchestrator consumer** (`bulkImportOrchestratorRunner.js`): HEADs each S3 file to get `ContentLength`, validates bucket against the allowlist (defense in depth), checks file size against min/max thresholds, and passes enriched inputs (with `fileSize`) to the event producer for byte-range fan-out. On S3 access errors or size violations the Task is set to `failed` with the error in `statusReason.text`. Added `Number.isFinite(fileSize)` guard after HEAD to reject missing `ContentLength`.
- **S3 NDJSON reader** (`s3NdjsonReader.js`): streams a byte range from S3 and yields parsed NDJSON lines. Handles partial lines at range boundaries (skips first partial line on non-zero start, reads one extra line past end). Validates `fileSize`, `byteRangeStart`, and `byteRangeEnd` before any network I/O.
- Added `databaseUpdateFactory` dependency to the orchestrator runner for Task status updates.

Jira: https://icanbwell.atlassian.net/browse/BAI-220

## Test plan
- [x] Orchestrator runner unit tests (6 tests): CloudEvent parsing, byte-range publishing, S3 error handling, Task not found, malformed messages
- [ ] Existing import tests pass in CI
- [ ] S3 NDJSON reader tests

### CODE REVIEW

File: src/operations/import/s3NdjsonReader.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            input: response.Body,
+            crlfDelay: Infinity
+        });
+
+        let lineNumber = 0;
+        let bytesRead = byteRangeStart;
+        let skippedFirst = isFirstRange;
+
+        for await (const line of rl) {
+            const lineBytes = Buffer.byteLength(line, 'utf8') + 1; // +1 for newline
```
Comment (mintusah25) [inline src/operations/import/s3NdjsonReader.js:103]: **[Medium] CRLF byte-accounting drift → silent duplication/loss.**

`lineBytes = Buffer.byteLength(line) + 1` hardcodes a 1-byte terminator, but `readline` is configured with `crlfDelay: Infinity`, which collapses `\r\n` (2 bytes on disk) into a single break. On a CRLF file, `bytesRead` undercounts by 1 byte per line, so the `bytesRead - lineBytes > byteRangeEnd` break fires *late* — this range over-reads past the real boundary, and those same lines get re-processed by the next range (whose first-line skip is drifted too). Result: **duplicate resource imports, no error raised**.

The tiling is correct for `\n`-only files, but nothing enforces that. Please either assert/document LF-only input, or derive the true terminator length (e.g. inspect the raw bytes rather than relying on `readline`).

File: src/operations/import/s3NdjsonReader.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            throw new Error(`Invalid byteRangeStart ${byteRangeStart} for "${filepath}"`);
+        }
+        if (!Number.isFinite(byteRangeEnd) || byteRangeEnd <= byteRangeStart) {
+            throw new Error(`Invalid byteRangeEnd ${byteRangeEnd} for "${filepath}"`);
+        }
+
+        const { bucket, key } = this.parseS3Uri(filepath);
+        const region = this.configManager.awsRegion || 'us-east-1';
+        const s3 = new S3({ region });
+        const maxLineSizeBytes = (this.configManager.bulkImportMaxLineSizeMb || 16) * 1024 * 1024;
```
Comment (mintusah25) [inline src/operations/import/s3NdjsonReader.js]: **[Low] Dead config knob.** `this.configManager.bulkImportMaxLineSizeMb` has no corresponding getter in `ConfigManager`, so it always evaluates to `undefined` and the `|| 16` fallback silently wins. The intended `BULK_IMPORT_MAX_LINE_SIZE_MB` env var is never read. Add the getter (mirroring the other `bulkImport*` getters), or hardcode 16 with a comment if it's not meant to be configurable.

File: src/operations/import/s3NdjsonReader.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        if (!Number.isFinite(byteRangeStart) || byteRangeStart < 0) {
+            throw new Error(`Invalid byteRangeStart ${byteRangeStart} for "${filepath}"`);
+        }
+        if (!Number.isFinite(byteRangeEnd) || byteRangeEnd <= byteRangeStart) {
+            throw new Error(`Invalid byteRangeEnd ${byteRangeEnd} for "${filepath}"`);
+        }
+
+        const { bucket, key } = this.parseS3Uri(filepath);
+        const region = this.configManager.awsRegion || 'us-east-1';
+        const s3 = new S3({ region });
```
Comment (mintusah25) [inline src/operations/import/s3NdjsonReader.js:72]: **[Minor] Reuse the existing S3 wrapper / avoid per-call client construction.** The repo already has `src/utils/s3Client.js` wrapping `@aws-sdk/client-s3` (and it configures endpoint/credentials for local/test). Here (and at `bulkImportOrchestratorRunner.js:109`) a fresh `new S3({ region })` is built on every range/message — no connection pooling and inconsistent with convention. Consider reusing the wrapper or at least caching the client on the instance.

Also — **`s3NdjsonReader` currently has no tests**, and the mock yields `[]`, so the consumer read loop is untested too. Given the boundary skip/over-read logic (and the CRLF issue above), please add unit tests: first/middle/last range, a line straddling a boundary, oversized line, invalid JSON, blank-line skipping — with both LF and CRLF fixtures.

File: src/utils/configManager.js [hunk 4]
```diff
@@ -1250,7 +1250,7 @@ class ConfigManager {
      */
     get bulkImportMinFileSizeMb() {
         const parsed = parseInt(env.BULK_IMPORT_MIN_FILE_SIZE_MB, 10);
-        return Number.isFinite(parsed) && parsed > 0 ? parsed : 50;
+        return Number.isFinite(parsed) && parsed >= 0 ? parsed : 0;
```
Comment (mintusah25) [inline src/utils/configManager.js]: **[Medium — confirm intent] Minimum-file-size protection is now off by default.** This changes the default from `50` / `> 0` to `0` / `>= 0`. Combined with the orchestrator's `if (minBytes > 0 && ...)` guard, the min-size check no longer runs unless explicitly configured — previously small files were rejected by default. If deliberate, fine, but it's an unflagged behavior change and the "file too small" test mocks the throw, so this config path isn't actually covered.

File: src/operations/import/bulkImportOrchestratorRunner.js [hunk 5]
```diff
@@ -89,9 +184,18 @@ class BulkImportOrchestratorRunner {
             return;
         }
 
+        let inputsWithSizes;
+        try {
+            inputsWithSizes = await this.headS3FilesAsync(inputs);
+        } catch (e) {
+            logError('S3 validation failed for import task', { taskId, error: e.message });
```
Comment (mintusah25) [inline src/operations/import/bulkImportOrchestratorRunner.js]: **[Medium] Transient S3 errors become permanent Task failures.** This `catch` marks the Task `failed` and returns (message consumed, no re-throw), so a throttle / `503` / timeout from the S3 HEAD is non-retryable — the whole import dies on a transient blip. Consider distinguishing transient (`SlowDown`, 5xx, timeouts) from permanent (`NoSuchKey`, `AccessDenied`): let transient errors throw so Kafka redelivers, and only mark `failed` on permanent ones.

File: src/tests/import/bulkImportOrchestratorRunner.test.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+describe('BulkImportOrchestratorRunner', () => {
+    let originalHeadS3FilesAsync;
+
+    beforeEach(async () => {
+        process.env.ENABLE_BULK_IMPORT = '1';
+        process.env.BULK_IMPORT_ALLOWED_S3_BUCKETS = 'allowed-bucket';
+        process.env.ENABLE_EVENTS_KAFKA_V2 = '1';
+        originalHeadS3FilesAsync = BulkImportOrchestratorRunner.prototype.headS3FilesAsync;
+        BulkImportOrchestratorRunner.prototype.headS3FilesAsync = async (inputs) =>
```
Comment (mintusah25) [inline src/tests/import/bulkImportOrchestratorRunner.test.js]: **[Coverage gap] The core new validation logic is never exercised.** Every test overrides `headS3FilesAsync` on the prototype here in `beforeEach`, so the actual S3 URI parsing, allowlist enforcement, `Number.isFinite`/`ContentLength` handling, and min/max size checks — the security-critical additions of this PR — have zero coverage. The tests only validate the *reaction* to a thrown error, not the logic that throws. Recommend adding direct unit tests for `headS3FilesAsync` with a mocked S3 client (allowed vs disallowed bucket, missing `ContentLength`, size over/under thresholds, malformed URI).

**Review by mintusah25 — COMMENTED** [review]
Solid, security-conscious work — the fail-closed bucket allowlist (empty ⇒ reject all), defense-in-depth validation in both `import.js` and the orchestrator, the `Number.isFinite(fileSize)` guard after HEAD, and wrapping the Kafka publish to prevent orphaned Tasks are all good. I verified the byte-range tiling is **correct for `\n`-delimited files**: the producer emits contiguous ranges (`end_i == start_{i+1}`), which is exactly what the reader's skip-first-partial + over-read logic requires.

A few things to address before merge (details inline):

- **[Medium] CRLF byte-accounting drift** in `s3NdjsonReader` can silently duplicate/lose records on `\r\n` files — assert/document LF-only or handle the real terminator length.
- **[Medium] Min-file-size default changed to 0** — the minimum-size check is now off by default; please confirm intent.
- **[Medium] Transient S3 errors become permanent Task failures** — no retry on throttle/5xx.
- **[Low] `bulkImportMaxLineSizeMb` has no getter** — the env knob is dead (`|| 16` masks it).
- **[Minor] Reuse the existing `s3Client.js` wrapper** instead of `new S3()` per call.
- **[Coverage] The core new logic is untested** — `headS3FilesAsync` is stubbed out in every orchestrator test, and `s3NdjsonReader` (the most bug-prone code) has no tests at all. The PR's own test-plan checkboxes for these are unchecked.

No crash-level blockers — DI wiring, `moment`, and the DB factories are all correctly available. Nice work overall; mainly want the CRLF assumption resolved and real tests for the reader + `headS3FilesAsync`.

File: src/utils/configManager.js (hunk 4 above)
Comment (adachaman-11) [reply src/utils/configManager.js]: `fhir- chauffeur` controls file sizing via repartition, server accepts any size when set to 0.

File: src/operations/import/bulkImportOrchestratorRunner.js (hunk 5 above)
Comment (adachaman-11) [reply src/operations/import/bulkImportOrchestratorRunner.js]: S3 logic is commented out. Will address when we uncomment in follow-up PR

File: src/tests/import/bulkImportOrchestratorRunner.test.js (hunk 6 above)
Comment (adachaman-11) [reply src/tests/import/bulkImportOrchestratorRunner.test.js]: Tests written but commented out — will uncomment alongside the logic in follow-up PR

File: src/operations/import/s3NdjsonReader.js (hunk 3 above)
Comment (adachaman-11) [reply src/operations/import/s3NdjsonReader.js:72]: S3 logic is commented out. Will address when we uncomment in follow-up PR

File: src/utils/configManager.js (hunk 4 above)
Comment (mintusah25) [reply src/utils/configManager.js]: No, FHIR Sever would reject any file having less then 50 MB. it is hard requirement.

**Review by mintusah25 — DISMISSED** [review]

**Review by mintusah25 — APPROVED** [review]
